src/Data/Store/Streaming.hs

```diff
@@ -86,91 +98,99 @@ encodeMessage (Message x) =
--- header, and then the actual 'Message'.
-peekMessage :: (MonadIO m, Store a) => ByteBuffer -> m (PeekMessage m a)
-peekMessage bb =
-   peekMessageHeader bb >>= \case
-        (Done (Message n)) -> peekSized bb n
-        NeedMoreInput _ ->
-            return $ NeedMoreInput (\ bs -> BB.copyByteString bb bs
-                                            >> peekMessage bb)
+-- | Decode some object from a 'ByteBuffer', by first reading its
+-- header, and then the actual data.
+peekMessage :: (MonadIO m, Store a) => FillByteBuffer i m -> ByteBuffer -> PeekMessage i m (Message a)
+peekMessage fill bb =
+  Message <$> (peekSized fill bb =<< peekMessageHeader fill bb)
 {-# INLINE peekMessage #-}
 
 -- | Decode a 'Message' from a 'ByteBuffer' and an action that can get
--- additional 'ByteString's to refill the buffer when necessary.
+-- additional inputs to refill the buffer when necessary.
 --
 -- The only conditions under which this function will give 'Nothing',
 -- is when the 'ByteBuffer' contains zero bytes, and refilling yields
 -- 'Nothing'.  If there is some data available, but not enough to
 -- decode the whole 'Message', a 'PeekException' will be thrown.
-decodeMessage :: (MonadIO m, Store a)
-            => ByteBuffer -> m (Maybe ByteString) -> m (Maybe (Message a))
-decodeMessage bb getBs =
-    decodeHeader bb getBs >>= \case
-        Nothing -> return Nothing
-        Just n -> decodeSized bb getBs n
-{-# INLINE decodeMessage #-}
-
-decodeHeader :: MonadIO m
-              => ByteBuffer
-              -> m (Maybe ByteString)
-              -> m (Maybe SizeTag)
-decodeHeader bb getBs =
-    peekMessageHeader bb >>= \case
-        (Done (Message n)) -> return (Just n)
-        (NeedMoreInput _) -> getBs >>= \case
-            Just bs -> BB.copyByteString bb bs >> decodeHeader bb getBs
-            Nothing -> BB.availableBytes bb >>= \case
-                0 -> return Nothing
-                n -> liftIO $ tooManyBytes headerLength n "Data.Store.Message.SizeTag"
-{-# INLINE decodeHeader #-}
-
-decodeSized :: (MonadIO m, Store a)
-            => ByteBuffer
-            -> m (Maybe ByteString)
-            -> Int
-            -> m (Maybe (Message a))
-decodeSized bb getBs n =
-    peekSized bb n >>= \case
-        Done message -> return (Just message)
-        NeedMoreInput _ -> getBs >>= \case
-            Just bs -> BB.copyByteString bb bs >> decodeSized bb getBs n
-            Nothing -> BB.availableBytes bb >>= \ available ->
-                liftIO $ tooManyBytes n available "Data.Store.Message.Message"
-{-# INLINE decodeSized #-}
+decodeMessage :: (Store a, MonadIO m) => FillByteBuffer i m -> ByteBuffer -> m (Maybe i) -> m (Maybe (Message a))
+decodeMessage fill bb getInp = runMaybeT $
+  iterTM (\consumeInp -> consumeInp =<< MaybeT getInp) (peekMessage fill bb) <|>
+  (do available <- BB.availableBytes bb
+      unless (available == 0) $
+        liftIO $ tooManyBytes available available "Data.Store.Message.Message"
```

Comment:
`liftIO $ tooManyBytes available available "Data.Store.Message.Message"` is wrong, the first argument to `tooManyBytes` should be the expected ones. I'll fix it soon.